haystack/components/evaluators/document_ndcg.py

```diff
@@ -0,0 +1,95 @@
+
+    Usage example:
+    ```python
+    from haystack import Document
+    from haystack.components.evaluators import DocumentNDCGEvaluator
+
+    evaluator = DocumentNDCGEvaluator()
+    result = evaluator.run(
+        ground_truth_documents=[[Document(content="France"), Document(content="Paris")]],
+        retrieved_documents=[[Document(content="France"), Document(content="Germany"), Document(content="Paris")]],
+    )
+    print(result["individual_scores"])
+    # [0.8869]
+    print(result["score"])
+    # 0.8869
+    ```
+    """
+
+    @component.output_types(score=float, individual_scores=List[float])
+    def run(
+        self, ground_truth_documents: List[List[Document]], retrieved_documents: List[List[Document]]
+    ) -> Dict[str, Any]:
+        """
+        Run the DocumentNDCGEvaluator on the given inputs.
+
+        `ground_truth_documents` and `retrieved_documents` must have the same length.
+        The list items within `ground_truth_documents` and `retrieved_documents` can differ in length.
+
+        :param ground_truth_documents:
+            Lists of expected documents, one list per question, either with relevance scores or sorted by relevance.
+        :param retrieved_documents:
+            Lists of retrieved documents, one list per question.
+        :returns:
+            A dictionary with the following outputs:
+            - `score` - The average of calculated scores.
+            - `individual_scores` - A list of numbers from 0.0 to 1.0 that represents the NDCG for each question.
+        """
+        if len(ground_truth_documents) != len(retrieved_documents):
+            msg = "The length of ground_truth_documents and retrieved_documents must be the same."
+            raise ValueError(msg)
+
+        for gt_docs in ground_truth_documents:
+            if any(doc.score is not None for doc in gt_docs) and any(doc.score is None for doc in gt_docs):
+                msg = "Either none or all documents in each list of ground_truth_documents must have a score."
+                raise ValueError(msg)
+
+        individual_scores = []
+
+        for gt_docs, ret_docs in zip(ground_truth_documents, retrieved_documents):
+            dcg = self._calculate_dcg(gt_docs, ret_docs)
+            idcg = self._calculate_idcg(gt_docs)
+            ndcg = dcg / idcg if idcg > 0 else 0
+            individual_scores.append(ndcg)
+
+        score = sum(individual_scores) / len(ground_truth_documents)
+
+        return {"score": score, "individual_scores": individual_scores}
+
+    def _calculate_dcg(self, gt_docs: List[Document], ret_docs: List[Document]) -> float:
+        dcg = 0.0
+        relevant_id_to_score = {doc.id: doc.score for doc in gt_docs}
+        for i, doc in enumerate(ret_docs):
+            if doc.id in relevant_id_to_score:  # TODO Related to https://github.com/deepset-ai/haystack/issues/8412
+                # If the gt document has a float score, use it; otherwise, use the inverse of the rank
```

Comment:
Why are we using the inverse of the rank as fallback? Effectively this would double the "rank-discount" of the retrieved document: One by doing dividing by (i +1) in line 85 and the other by dividing by log2(i + 2) in line 86.

I guess a better fallback would be to just use value `1` which would translate into a simple binary relevance schema according to https://en.wikipedia.org/wiki/Discounted_cumulative_gain

Reply:
@julian-risch yes, I'm using [HotPot QA dataset](https://huggingface.co/datasets/hotpotqa/hotpot_qa) from hugging face and it doesn't provide scores.

Reply:
If we change the fallback to binary relevance, what you could do is calculate scores yourself before passing the documents to the DocumentNDCGEvaluator. For example:
```python
for i, doc in enumerate(docs, 1):
    doc.score = 1 / i
```
That would work for you too right?

Reply:
@julian-risch I can understand the intuition now. Still I'd probably not make this the default behavior: When supplying ground-truth docs, I wouldn't expect that the order of them makes a difference, tbh.
And if you really need that, you could simply pass scores, no?

Anyways, I think there is an error in the implemetatoin of the intuiton. If I got it correct, then document relevance should be based on the order of the passed ground-truth docs. In the current implementation it's instead based on the order of the retrieved documents: `relevance = 1 / (i + 1)` where `i` is the index of the retrieved doc, not the ground-truth doc.

Reply:
Yes, true. I will change the fallback to binary relevance. 👍